### DISCUSSION TITLE
Alert Regex

### DISCUSSION THREAD

**Discussion by shoot991** [nextgenhealthcare/connect#5533] [Q&A] [answered]
Hello,

I need help getting a regular expression to work on the alerting page. I am using Mirth 3.12.0
The problem is some kind of permissions on a SFTP server that causes this error even though the file has been deleted.
I would like to block those alerts.

The expression I have is:

/^((?!Error deleting file from)[\s\S])*$/g

What I am trying to do with this is block alerting for just this error.

Error deleting file from [/example/file.pdf]​

[2022-11-07 09:00:16,465] ERROR (com.mirth.connect.connectors.file.FileReceiver:27 ): Error processing file in channel: 16740753-f291-4042-9d95-0cffde4fb17b
com.mirth.connect.connectors.file.FileConnectorException: Error deleting file from [/example/file.pdf]
at com.mirth.connect.connectors.file.FileReceiver.processFile(FileReceiver.java:490)
at com.mirth.connect.connectors.file.FileReceiver.processFiles(FileReceiver.java:328)
at com.mirth.connect.connectors.file.FileReceiver.pol l(FileReceiver.java:236)
at com.mirth.connect.donkey.server.channel.PollConnectorJob.execute(PollConnectorJob.java:49)
at org.quartz.core.JobRunShell.run(JobRunShell.java:2 13)
at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557)​
Tags: None

![image](https://user-images.githubusercontent.com/1590974/202031189-e16fcbfc-d4e1-46a4-9bd8-991830f4e547.png)

**Comment by jonbartels** [comment, marked as answer]
Try the regex at https://regex101.com/r/lONYpq/1. You can click "save" to save a new version as you make edits and need more help.

The regex seems to work. If I change the error from "Error deleting file" to "Error reading file" the regex matches.

Can you post at least two sample alerts for:
* message that SHOULD match the regex and alert
* message that SHOULD NOT match the regex and not alert (I think your original post has this)

Please use code tags when you post those messages.

**Reply by shoot991** [reply]
Thank you for your quick reply. I am trying to get the full error message. I might have to force an error somehow.

Here is the error from the email that should pass: 
`java.net.ConnectException: Connection refused: connect`

Here is the error from an email that should not pass:
`Error deleting file from [/outbound/folder/censored.pdf]`

I did notice you are using Java 8 as the Regex format. Did I just enter it wrong in Mirth?

**Reply by shoot991** [reply]
Those are ${errorMessage} from the Alert

**Reply by jonbartels** [reply]
Does this version show the expected inputs for `errorMessage` and the correct matches for alerting? https://regex101.com/r/lONYpq/3

The relevant manual page is https://docs.nextgen.com/bundle/Mirth_User_Guide_41/page/connect/connect/topics/c_Alert_Error_Types_and_Regex_connect_ug.html

The code https://github.com/nextgenhealthcare/connect/blob/e66d1122dbeebfcf5a8939beb93d720b2d5071fb/server/src/com/mirth/connect/server/alert/DefaultAlertWorker.java#L114 shows that MC is going to match on a formatted error message built by https://github.com/nextgenhealthcare/connect/blob/e66d1122dbeebfcf5a8939beb93d720b2d5071fb/server/src/com/mirth/connect/util/ErrorMessageBuilder.java#L19

The problem here is something like a regex matching on multiple lines is always going to have a condition where ONE of those lines DOES NOT contain the string you want to exclude.

**Reply by shoot991** [reply]
Got it. 
"always going to have a condition where ONE of those lines DOES NOT contain the string you want to exclude" That makes a lot of sense as to why the message passes.

I am going to do some digging to see if there is another way for me to do this. Thank you very much for all of your help!

**Reply by jonbartels** [reply]
Try this `(?s)^(?!.*\bError deleting file from[^$\n]*(?:\n|$)).*` at https://regex101.com/r/lONYpq/4

I grabbed it from - https://stackoverflow.com/questions/37138155/regex-multiline-match-only-if-all-lines-do-not-contain . I will admit I do not really understand how it works but it passes some simple test cases.

I tested it with:
* Your original full stack trace, it does not match. This is the desired outcome
* By changing `Error deleting file from` to `Could not read` and it matches. This would alert.

What if you tried to prevent the error in the first place? Is this error coming from an SFTP source connector or destination connector? I wonder if a response transformer would catch it.
